pkg/capacityscheduling/capacity_scheduling.go

```diff
@@ -0,0 +1,841 @@
+	var resultLock sync.Mutex
+	var candidates []defaultpreemption.Candidate
+	checkNode := func(i int) {
+		nodeInfoCopy := potentialNodes[i].Clone()
+		stateCopy := state.Clone()
+
+		pods, numPDBViolations, fits := selectVictimsOnNode(ctx, fh, stateCopy, pod, nodeInfoCopy, pdbs)
+		if fits {
+			resultLock.Lock()
+			victims := extenderv1.Victims{
+				Pods:             pods,
+				NumPDBViolations: int64(numPDBViolations),
+			}
+			c := candidate{
+				victims: &victims,
+				name:    nodeInfoCopy.Node().Name,
+			}
+			candidates = append(candidates, &c)
+			resultLock.Unlock()
+		}
+	}
+	pluginsutil.Until(ctx, len(potentialNodes), checkNode)
+	return candidates
+}
+
+// nodesWherePreemptionMightHelp returns a list of nodes with failed predicates
+// that may be satisfied by removing pods from the node.
+func nodesWherePreemptionMightHelp(nodes []*framework.NodeInfo, m framework.NodeToStatusMap) []*framework.NodeInfo {
+	var potentialNodes []*framework.NodeInfo
+	for _, node := range nodes {
+		name := node.Node().Name
+		// We reply on the status by each plugin - 'Unschedulable' or 'UnschedulableAndUnresolvable'
+		// to determine whether preemption may help or not on the node.
+		if m[name].Code() == framework.UnschedulableAndUnresolvable {
+			continue
+		}
+		potentialNodes = append(potentialNodes, node)
+	}
+	return potentialNodes
+}
+
+func selectVictimsOnNode(
+	ctx context.Context,
+	ph framework.PreemptHandle,
+	state *framework.CycleState,
+	pod *v1.Pod,
+	nodeInfo *framework.NodeInfo,
+	pdbs []*policy.PodDisruptionBudget,
+) ([]*v1.Pod, int, bool) {
+	elasticQuotaSnapshotState, err := getElasticQuotaSnapshotState(state)
+	if err != nil {
+		klog.Errorf("error reading %q from cycleState: %v", ElasticQuotaSnapshotKey, err)
+		return nil, 0, false
+	}
+
+	preFilterState, err := getPreFilterState(state)
+	if err != nil {
+		klog.Errorf("error reading %q from cycleState: %v", preFilterStateKey, err)
+		return nil, 0, false
+	}
+
+	var potentialVictims []*v1.Pod
+	var moreThanMinWithPreemptor bool
+
+	removePod := func(rp *v1.Pod) error {
+		if err := nodeInfo.RemovePod(rp); err != nil {
+			return err
+		}
+		status := ph.RunPreFilterExtensionRemovePod(ctx, state, pod, rp, nodeInfo)
+		if !status.IsSuccess() {
+			return status.AsError()
+		}
+		return nil
+	}
+
+	addPod := func(ap *v1.Pod) error {
+		nodeInfo.AddPod(ap)
+		status := ph.RunPreFilterExtensionAddPod(ctx, state, pod, ap, nodeInfo)
+		if !status.IsSuccess() {
+			return status.AsError()
+		}
+		return nil
+	}
+
+	podPriority := podutil.GetPodPriority(pod)
+	preemptorElasticQuotaInfo, preemptorWithElasticQuota := elasticQuotaSnapshotState.elasticQuotaInfos[pod.Namespace]
+
+	// Check if there is elastic quota in the preemptor's namespace.
+	if preemptorWithElasticQuota {
+		moreThanMinWithPreemptor = preemptorElasticQuotaInfo.overUsed(preFilterState.Resource, preemptorElasticQuotaInfo.Min, resourceAdd)
```

Comment:
I think it's `moreThanMaxWithPreemptor`, not more than min.

- If it's more than max, for sure we should preempt pods belonging to the same EQ.
    - If preempting pods from the same EQ works, we're done.
    - If not, it indicates we may need to **additionally** preempt pods belongs to other EQs that have over-used their min.
- If it's not, then pods that belong to an EQ whose usage > min are potential victims.

WDYT?

Reply:
I think that "moreThanMinWithPreemptor" makes more sense. 

- If it is more than min, we have to preempt pods belonging to the same EQ. We can not preempt pods belongs to other EQs. Because those extra resources belong to best-effort resources which can not be guaranteed. So, we can not preempt pods from other EQs

-  If it is less than min, it means some other EQs use more resources than its guaranteed resources which they need to return back.

Using "moreThanMaxWithPreemptor" might cause more issues, for example, if one user asks for some resources which cause totalResource > min, but < max. At this time, I think that we can only preempt the pods from the same EQ. We should not preempt pods from other EQ even if they use more resources than min.

Reply:
> * If it is more than min, we have to preempt pods belonging to the same EQ. We can not preempt pods belongs to other EQs. Because those extra resources belong to best-effort resources which can not be guaranteed. So, we can not preempt pods from other EQs

What if preempting pods within the EQ doesn't help? Is it legal to preempt over-used resources from other EQs if the current pod has a higher priority? If not, it seems low-priority pods in other EQs that comes first can reserve extra resource without worrying about being preempted. Which doesn't look correct to me.

So my point is that among EQs that over-used resources, Pod's priority is the key to determine whether it can preempt others.

Reply:
Exactly. I totally agree with you. Here is what I mean: 

- request resource + allocated resource <= min, then preempt from other EQs

- request resource + allocated resource > max, then preempt from the same EQ

- min < request resource + allocated resource < max, preempt based on the priority. It can preempt pods from the same EQ or from the other EQs whose total resource > min.

Reply:
This is actually divided into two layers:  EQ and Priority Class.
I think it's multi tenant. The Priority Class in the same EQ or tenant is comparable. The priority Class in the different EQ or tenant isn't comparable.
The extra resource is first come, first served. If min < request resource + allocated resource < max, the preempt only allowed to preempt the pods from the same EQ.

Reply:
> I think it's multi tenant. The Priority Class in the same EQ or tenant is comparable. The priority Class in the different EQ or tenant isn't comparable.
A little more: 
Different users(tenants) have different ways to use priority Class. tenant A: high(100) low(10) tenant B: high(1000) low(100). So The priority Class in the different EQ or tenant isn't comparable.

Reply:
> The extra resource is first come, first served. If min < request resource + allocated resource < max, the preempt only allowed to preempt the pods from the same EQ.

I'm afraid this behavior is not mentioned in the KEP. I'm not sure it's the right way as it violates the API semantics of Priority Class. Could you submit a PR describing this behavior in the KEP's [story section](https://github.com/kubernetes-sigs/scheduler-plugins/tree/master/kep/9-capacity-scheduling#story-1)? and I may bring it up to the community meeting (next Thursday) for discussion.

Reply:
> Different users(tenants) have different ways to use priority Class. tenant A: high(100) low(10) tenant B: high(1000) low(100). So The priority Class in the different EQ or tenant isn't comparable.

This would surprise users IMHO. Is it possible to make it a plugin option? I'm fine with making it default.

Reply:
After offline discussion, we decided that the current version process is as follows to make sure that they are used in the same way as yarn users.
1. When quota.used + preemptor.request < min, victims will be selected from the EQ which used > min
2. When quota.used + preemptor.request > min, victims will be selected from the same EQ.
@Huang-Wei @gongx

Reply:
@denkensk can you add these 2 cases into the KEP's story sections. Thanks!

BTW: a nit on (1), `<` should be `<=` :)

Reply:
Sure I will update it into KEP in another pr,